velox/functions/lib/SubscriptUtil.cpp

```diff
@@ -14,10 +14,258 @@
+
+    // Handle NULLs.
+    if (!found) {
+      nullsBuilder.setNull(row);
+    }
+  };
+
+  // When second argument ("at") is a constant.
+  if (decodedIndices->isConstantMapping()) {
+    auto searchKey = decodedIndices->valueAt<TKey>(0);
+    rows.applyToSelected(
+        [&](vector_size_t row) { processRow(row, searchKey); });
+  }
+
+  // When the second argument ("at") is also a variable vector.
+  else {
+    rows.applyToSelected([&](vector_size_t row) {
+      auto searchKey = decodedIndices->valueAt<TKey>(row);
+      processRow(row, searchKey);
+    });
+  }
+
+  // Subscript into empty maps always returns NULLs. Check added at the end to
+  // ensure user error checks for indices are not skipped.
+  if (baseMap->mapValues()->size() == 0) {
+    return BaseVector::createNullConstant(
+        baseMap->mapValues()->type(), rows.end(), context.pool());
+  }
+
+  return BaseVector::wrapInDictionary(
+      nullsBuilder.build(), indices, rows.end(), baseMap->mapValues());
+}
+
+VectorPtr applyMapComplexType(
+    const SelectivityVector& rows,
+    const VectorPtr& mapArg,
+    const VectorPtr& indexArg,
+    exec::EvalCtx& context) {
+  auto* pool = context.pool();
+
+  // Use indices with the mapValues wrapped in a dictionary vector.
+  BufferPtr indices = allocateIndices(rows.end(), pool);
+  auto rawIndices = indices->asMutable<vector_size_t>();
+
+  // Create nulls for lazy initialization.
+  NullsBuilder nullsBuilder(rows.end(), pool);
+
+  // Get base MapVector.
+  exec::LocalDecodedVector mapHolder(context, *mapArg, rows);
+  auto decodedMap = mapHolder.get();
+  auto baseMap = decodedMap->base()->as<MapVector>();
+  auto mapIndices = decodedMap->indices();
+
+  // Get map keys.
+  auto mapKeys = baseMap->mapKeys();
+  exec::LocalSelectivityVector allElementRows(context, mapKeys->size());
+  allElementRows->setAll();
+  exec::LocalDecodedVector mapKeysHolder(context, *mapKeys, *allElementRows);
+  auto mapKeysDecoded = mapKeysHolder.get();
+  auto mapKeysBase = mapKeysDecoded->base();
+  auto mapKeysIndices = mapKeysDecoded->indices();
+
+  // Get index vector (second argument).
+  exec::LocalDecodedVector indexHolder(context, *indexArg, rows);
+  auto decodedIndices = indexHolder.get();
+  auto searchBase = decodedIndices->base();
+  auto searchIndices = decodedIndices->indices();
+
+  auto rawSizes = baseMap->rawSizes();
+  auto rawOffsets = baseMap->rawOffsets();
+
+  // Search the key in each row.
+  rows.applyToSelected([&](vector_size_t row) {
+    size_t mapIndex = mapIndices[row];
+    size_t size = rawSizes[mapIndex];
+    size_t offset = rawOffsets[mapIndex];
+
+    bool found = false;
+    auto searchIndex = searchIndices[row];
+    for (auto i = 0; i < size; i++) {
+      if (mapKeysBase->equalValueAt(
+              searchBase, mapKeysIndices[offset + i], searchIndex)) {
+        rawIndices[row] = offset + i;
+        found = true;
+        break;
+      }
+    }
+
+    if (!found) {
+      nullsBuilder.setNull(row);
```

Comment:
Hi @laithsakka, could you please share why the result is set to `NULL` here when the key is not found in map? Could we throw an error here instead?